// @frontegg/redux-store/node/auth/reducer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.reducer = exports.actions = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _toolkit = require("@reduxjs/toolkit");

var _utils = require("./utils");

var _initialState = require("./initialState");

var _constants = require("../constants");

var _LoginState = require("./LoginState");

var _ActivateState = require("./ActivateState");

var _AcceptInvitationState = require("./AcceptInvitationState");

var _ForgotPasswordState = require("./ForgotPasswordState");

var _SSOState = require("./SSOState");

var _MfaState = require("./MfaState");

var _ProfileState = require("./ProfileState");

var _TeamState = require("./TeamState");

var _SocialLogins = require("./SocialLogins");

var _SignUp = require("./SignUp");

var _ApiTokensState = require("./ApiTokensState");

var _SecurityPolicyState = require("./SecurityPolicyState");

var _AccountSettingsState = require("./AccountSettingsState");

var _TenantsState = require("./TenantsState");

var _RolesState = require("./RolesState");

var _ResetPhoneNumberState = require("./ResetPhoneNumberState");

var _SessionsState = require("./SessionsState");

var _SessionsPolicyState = require("./SessionsPolicyState");

var _RestrictionsState = require("./RestrictionsState");

var _Provisioning = require("./Provisioning");

var _ImpersonationState = require("./ImpersonationState");

const {
  reducer,
  actions: sliceActions
} = (0, _toolkit.createSlice)({
  name: _constants.authStoreName,
  initialState: _initialState.initialState,
  reducers: (0, _extends2.default)({
    resetState: state => (0, _extends2.default)({}, state, _initialState.reinitializeState),
    setState: (state, {
      payload
    }) => (0, _extends2.default)({}, state, payload),
    setUser: (0, _utils.typeReducer)('user')
  }, _LoginState.loginReducers, _SocialLogins.socialLoginsReducer, _ActivateState.activateAccountReducers, _ImpersonationState.impersonateReducers, _AcceptInvitationState.acceptInvitationReducers, _ForgotPasswordState.forgotPasswordReducers, _ResetPhoneNumberState.resetPhoneNumberReducers, _SignUp.signUpReducers, _ProfileState.profileReducers, _SSOState.ssoReducers, _MfaState.mfaReducers, _TeamState.teamReducers, _ApiTokensState.apiTokensReducers, _SecurityPolicyState.securityPolicyReducers, _AccountSettingsState.accountSettingsReducers, _TenantsState.tenantsReducers, _RolesState.rolesReducers, _SessionsState.sessionsReducers, _SessionsPolicyState.sessionsPolicyReducers, _RestrictionsState.restrictionsReducers, _Provisioning.provisioningReducers)
});
exports.reducer = reducer;
const actions = (0, _extends2.default)({}, sliceActions, _LoginState.loginActions, _SocialLogins.socialLoginsActions, _ActivateState.activateAccountActions, _AcceptInvitationState.acceptInvitationActions, _ForgotPasswordState.forgotPasswordActions, _ResetPhoneNumberState.resetPhoneNumberActions, _SignUp.signUpActions, _ProfileState.profileActions, _SSOState.ssoActions, _MfaState.mfaActions, _TeamState.teamActions, _ApiTokensState.apiTokensActions, _SecurityPolicyState.securityPolicyActions, _AccountSettingsState.accountSettingsActions, _TenantsState.tenantsActions, _RolesState.rolesActions, _SessionsState.sessionsActions, _SessionsPolicyState.sessionsPolicyActions, _RestrictionsState.restrictionsActions, _Provisioning.provisioningActions, _ImpersonationState.impersonateActions);
exports.actions = actions;
