// @frontegg/redux-store/node/auth/TeamState/saga.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.teamSagas = teamSagas;
exports.teamSagasMock = teamSagasMock;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _effects = require("redux-saga/effects");

var _restApi = require("@frontegg/rest-api");

var _reducer = require("../reducer");

var _interfaces = require("./interfaces");

var _constants = require("../../constants");

var _utils = require("../utils");

var _dummy = require("../dummy");

var _uuid = require("uuid");

const _excluded = ["callback"],
      _excluded2 = ["roles"],
      _excluded3 = ["callback"],
      _excluded4 = ["callback", "profileImage"],
      _excluded5 = ["callback"],
      _excluded6 = ["callback", "userId"],
      _excluded7 = ["callback"],
      _excluded8 = ["callback"],
      _excluded9 = ["callback"],
      _excluded10 = ["callback"],
      _excluded11 = ["callback"],
      _excluded12 = ["callback"],
      _excluded13 = ["callback"],
      _excluded14 = ["callback", "profileImage"],
      _excluded15 = ["callback"],
      _excluded16 = ["callback"],
      _excluded17 = ["callback"],
      _excluded18 = ["callback"],
      _excluded19 = ["callback"],
      _excluded20 = ["callback"];

const selectTeamState = () => (0, _effects.select)(_ => _[_constants.authStoreName].teamState); // eslint-disable-next-line @typescript-eslint/no-explicit-any


function* loadUsers({
  payload
}) {
  var _payload$pageSize, _payload$pageOffset, _payload$filter, _payload$sort;

  const {
    silentLoading,
    callback
  } = payload;
  const teamState = yield selectTeamState();
  const pageSize = (_payload$pageSize = payload.pageSize) != null ? _payload$pageSize : teamState.pageSize;
  const pageOffset = (_payload$pageOffset = payload.pageOffset) != null ? _payload$pageOffset : teamState.pageOffset;
  const filter = (_payload$filter = payload.filter) != null ? _payload$filter : teamState.filter;
  const sort = (_payload$sort = payload.sort) != null ? _payload$sort : teamState.sort;
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.USERS,
    value: !silentLoading
  }));
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    pageSize,
    pageOffset,
    filter,
    sort
  }));

  try {
    const [{
      items: users,
      totalPages,
      totalItems
    }, {
      items: roles
    }, {
      items: permissions
    }] = yield (0, _effects.all)([(0, _effects.call)(_restApi.api.teams.loadUsers, {
      pageSize,
      pageOffset,
      filter,
      sort
    }), (0, _effects.call)(_restApi.api.teams.loadAvailableRoles), (0, _effects.call)(_restApi.api.teams.loadAvailablePermissions)]);
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      users,
      totalPages,
      totalItems,
      roles,
      permissions
    }));
    callback == null ? void 0 : callback(users);
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setTeamError({
      key: _interfaces.TeamStateKeys.USERS,
      value: e.message
    }));
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      totalPages: 0,
      users: []
    }));
    callback == null ? void 0 : callback(null, e);
  }

  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.USERS,
    value: false
  }));
} // eslint-disable-next-line @typescript-eslint/no-explicit-any


function* loadAllSubTenantsUsers({
  payload
}) {
  var _payload$_limit, _payload$_offset, _payload$_filter, _payload$_sortBy, _payload$_order;

  const {
    silentLoading,
    callback
  } = payload;
  const teamState = yield selectTeamState();

  const _limit = (_payload$_limit = payload._limit) != null ? _payload$_limit : teamState.allUsersQueryParams._limit;

  const _offset = (_payload$_offset = payload._offset) != null ? _payload$_offset : teamState.allUsersQueryParams._offset;

  const _filter = (_payload$_filter = payload._filter) != null ? _payload$_filter : teamState.allUsersQueryParams._filter;

  const _sortBy = (_payload$_sortBy = payload._sortBy) != null ? _payload$_sortBy : teamState.allUsersQueryParams._sortBy;

  const _order = (_payload$_order = payload._order) != null ? _payload$_order : teamState.allUsersQueryParams._order;

  const allUsersQueryParams = {
    _limit: _limit || 20,
    _offset: _offset || 0,
    _filter: _filter || '',
    _sortBy: _sortBy || 'name',
    _order: _order || 'DESC'
  };
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.USERS,
    value: !silentLoading
  }));
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    allUsersQueryParams
  }));

  try {
    const [{
      items: users,
      _metadata: {
        totalPages,
        totalItems
      }
    }, {
      items: roles
    }, {
      items: permissions
    }] = yield (0, _effects.all)([(0, _effects.call)(_restApi.api.subTenants.loadAllUsers, (0, _extends2.default)({}, allUsersQueryParams)), (0, _effects.call)(_restApi.api.teams.loadAvailableRoles), (0, _effects.call)(_restApi.api.teams.loadAvailablePermissions)]);
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      allUsers: users,
      totalPages,
      totalItems,
      roles,
      permissions
    }));
    callback == null ? void 0 : callback(users);
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setTeamError({
      key: _interfaces.TeamStateKeys.USERS,
      value: e.message
    }));
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      totalPages: 0,
      users: []
    }));
    callback == null ? void 0 : callback(null, e);
  }

  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.USERS,
    value: false
  }));
}

function* loadRoles({
  payload
}) {
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.ROLES_AND_PERMISSIONS,
    value: true
  }));

  try {
    var _payload$callback;

    const [{
      items: roles
    }, {
      items: permissions
    }] = yield (0, _effects.all)([(0, _effects.call)(_restApi.api.teams.loadAvailableRoles), (0, _effects.call)(_restApi.api.teams.loadAvailablePermissions)]);
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      roles,
      permissions
    }));
    payload == null ? void 0 : (_payload$callback = payload.callback) == null ? void 0 : _payload$callback.call(payload, {
      roles,
      permissions
    });
  } catch (e) {
    var _payload$callback2;

    payload == null ? void 0 : (_payload$callback2 = payload.callback) == null ? void 0 : _payload$callback2.call(payload, null, e);
    yield (0, _effects.put)(_reducer.actions.setTeamError({
      key: _interfaces.TeamStateKeys.ROLES_AND_PERMISSIONS,
      value: e.message
    }));
  }

  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.ROLES_AND_PERMISSIONS,
    value: true
  }));
}

function* addUser({
  payload
}) {
  const {
    callback
  } = payload,
        body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded);
  const teamState = yield selectTeamState();
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    addUserDialogState: (0, _extends2.default)({}, teamState.addUserDialogState, {
      loading: true
    })
  }));

  try {
    const res = yield (0, _effects.call)(_restApi.api.teams.addUser, body);
    const {
      roles
    } = res,
          userWithoutRoleIds = (0, _objectWithoutPropertiesLoose2.default)(res, _excluded2);
    const roleIds = roles.map(role => role.id);
    const newUser = (0, _extends2.default)({}, userWithoutRoleIds, {
      roleIds
    });
    callback == null ? void 0 : callback(newUser);
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      users: [newUser, ...teamState.users],
      addUserDialogState: {
        open: false,
        loading: false
      }
    }));
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      addUserDialogState: (0, _extends2.default)({}, teamState.addUserDialogState, {
        loading: false,
        error: e.message
      })
    }));
    callback == null ? void 0 : callback(null, e.message);
  }
}

function* addUsersBulk({
  payload
}) {
  const {
    callback
  } = payload;
  const teamState = yield selectTeamState();
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    addUserDialogState: (0, _extends2.default)({}, teamState.addUserDialogState, {
      loading: true
    })
  }));
  const allowedEmails = [];
  const unallowedEmails = [];
  const bodies = payload.emails.map(email => ({
    email,
    roleIds: payload.roleIds
  }));
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    addUserDialogState: {
      open: true,
      loading: true
    }
  }));

  for (let i = 0; i < bodies.length; i++) {
    const body = bodies.at(i);

    try {
      yield (0, _effects.call)(_restApi.api.teams.addUser, body);
      allowedEmails.push(body.email);
    } catch (e) {
      unallowedEmails.push(body.email);
    }
  }

  const queryObject = {
    pageOffset: 0,
    pageSize: 10,
    filter: [],
    silentLoading: payload.emails.length > 0
  };
  yield (0, _effects.put)(_reducer.actions.loadUsers(queryObject));
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    addUserDialogState: {
      loading: false
    }
  }));
  callback == null ? void 0 : callback({
    unallowedEmails,
    allowedEmails
  });
}

function* addUserToSubTenants({
  payload
}) {
  const {
    callback
  } = payload,
        body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded3);
  const teamState = yield selectTeamState();
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    addUserDialogState: (0, _extends2.default)({}, teamState.addUserDialogState, {
      loading: true
    })
  }));

  try {
    yield (0, _effects.call)(_restApi.api.subTenants.addUserToTenantAndSubTenants, body);
    const {
      items: users,
      _metadata: {
        totalPages,
        totalItems
      }
    } = yield (0, _effects.call)(_restApi.api.subTenants.loadAllUsers, {
      _limit: 20,
      _offset: 0,
      _filter: '',
      _sortBy: 'name',
      _order: 'DESC'
    });
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      allUsers: users,
      totalPages,
      totalItems,
      addUserDialogState: {
        open: false,
        loading: false
      }
    }));
    callback == null ? void 0 : callback(null);
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      addUserDialogState: (0, _extends2.default)({}, teamState.addUserDialogState, {
        loading: false,
        error: e.message
      })
    }));
    callback == null ? void 0 : callback(null, e.message);
  }
}

function* updateUser({
  payload
}) {
  const {
    callback
  } = payload,
        body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded4);
  const {
    id: userId
  } = body;
  const teamState = yield selectTeamState();
  const oldUserData = teamState.users.find(user => user.id === body.id);
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.UPDATE_USER,
    value: userId || ''
  }));
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    addUserDialogState: (0, _extends2.default)({}, teamState.addUserDialogState, {
      loading: true
    })
  }));
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    addUserDialogState: (0, _extends2.default)({}, teamState.addUserDialogState, {
      loading: true
    }),
    users: teamState.users.map(user => {
      if (user.id === body.id) {
        return (0, _extends2.default)({}, user, body);
      }

      return user;
    })
  }));

  try {
    var _body$roleIds;

    if (oldUserData.roleIds.length > 0 && ((_body$roleIds = body.roleIds) == null ? void 0 : _body$roleIds.length) === 0) {
      body.roleIds = [''];
    }

    const {
      item: newUser
    } = yield (0, _effects.call)(_restApi.api.teams.updateUser, body);
    callback == null ? void 0 : callback(newUser);
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      users: teamState.users.map(user => user.id === newUser.id ? (0, _extends2.default)({}, user, newUser, {
        createdAt: user.createdAt,
        customData: user.customData,
        lastLogin: user.lastLogin
      }) : user)
    }));
    yield (0, _effects.put)(_reducer.actions.setTeamLoader({
      key: _interfaces.TeamStateKeys.UPDATE_USER,
      value: false
    }));
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      addUserDialogState: (0, _extends2.default)({}, teamState.addUserDialogState, {
        loading: false,
        error: e.message
      }),
      users: teamState.users.map(user => user.id === body.id ? (0, _extends2.default)({}, user, oldUserData) : user)
    }));
    yield (0, _effects.put)(_reducer.actions.setTeamLoader({
      key: _interfaces.TeamStateKeys.UPDATE_USER,
      value: false
    }));
    callback == null ? void 0 : callback(null, e.message);
  }
}

function* deleteUser({
  payload
}) {
  const {
    callback
  } = payload,
        body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded5);
  const teamState = yield selectTeamState();
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    deleteUserDialogState: (0, _extends2.default)({}, teamState.deleteUserDialogState, {
      loading: true
    })
  }));

  try {
    yield (0, _effects.call)(_restApi.api.teams.deleteUser, body);
    callback == null ? void 0 : callback(true);
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      users: teamState.users.filter(user => user.id !== body.userId),
      deleteUserDialogState: {
        open: false,
        loading: false
      }
    }));
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      deleteUserDialogState: (0, _extends2.default)({}, teamState.deleteUserDialogState, {
        loading: false,
        error: e.message
      })
    }));
    callback == null ? void 0 : callback(null, e.message);
  }
}

function* setUserRolesForSubTenants({
  payload
}) {
  const {
    callback,
    userId
  } = payload,
        body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded6);
  const teamState = yield selectTeamState();
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    addUserDialogState: (0, _extends2.default)({}, teamState.addUserDialogState, {
      loading: true
    })
  }));

  try {
    yield (0, _effects.call)(_restApi.api.subTenants.setUserRolesForSubTenants, userId, body);
    const updatedUser = teamState.allUsers.find(user => user.id === userId);
    let newTenants = [];

    if (updatedUser) {
      newTenants = updatedUser.tenants.map(tenant => {
        var _body$subTenantsRoles;

        return (0, _extends2.default)({}, tenant, {
          roles: ((_body$subTenantsRoles = body.subTenantsRoles.find(roleUpdate => roleUpdate.tenantId === tenant.tenantId)) == null ? void 0 : _body$subTenantsRoles.roleIds.map(roleId => {
            const role = teamState.roles.find(({
              id
            }) => roleId === id);
            return role;
          }).filter(role => role)) || tenant.roles
        });
      });
    }

    yield (0, _effects.put)(_reducer.actions.setTeamState({
      allUsers: [...teamState.allUsers.filter(user => user.id !== userId), ...(updatedUser ? [(0, _extends2.default)({}, updatedUser, {
        tenants: newTenants
      })] : [])],
      addUserDialogState: {
        open: false,
        loading: false
      }
    }));
    callback == null ? void 0 : callback(true);
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      addUserDialogState: (0, _extends2.default)({}, teamState.addUserDialogState, {
        loading: false,
        error: e.message
      })
    }));
    callback == null ? void 0 : callback(null, e.message);
  }
}

function* deleteUserFromSubTenants({
  payload
}) {
  const {
    callback
  } = payload,
        body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded7);
  const teamState = yield selectTeamState();
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    deleteUserDialogState: (0, _extends2.default)({}, teamState.deleteUserDialogState, {
      loading: true
    })
  }));

  try {
    yield (0, _effects.call)(_restApi.api.subTenants.removeUserFromTenantAndSubTenants, body);
    callback == null ? void 0 : callback(true);
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      allUsers: teamState.allUsers.filter(user => user.id !== body.userId && user.tenants.length === body.subTenants.length),
      deleteUserDialogState: {
        open: false,
        loading: false
      }
    }));
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      deleteUserDialogState: (0, _extends2.default)({}, teamState.deleteUserDialogState, {
        loading: false,
        error: e.message
      })
    }));
    callback == null ? void 0 : callback(null, e.message);
  }
}

function* resendActivationLink({
  payload
}) {
  const {
    callback
  } = payload,
        body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded8);
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.RESEND_ACTIVATE_LINK,
    value: body.userId
  }));

  try {
    yield (0, _effects.call)(_restApi.api.teams.resendActivationLink, body);
    callback == null ? void 0 : callback(true);
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setTeamError({
      key: _interfaces.TeamStateKeys.RESEND_ACTIVATE_LINK,
      value: e.message
    }));
    callback == null ? void 0 : callback(null, e.message);
  }

  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.RESEND_ACTIVATE_LINK,
    value: false
  }));
}

function* resendInvitationLink({
  payload
}) {
  const {
    callback
  } = payload,
        body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded9);
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.RESEND_INVITATION_LINK,
    value: body.email
  }));

  try {
    yield (0, _effects.call)(_restApi.api.teams.resendInvitationLink, body);
    callback == null ? void 0 : callback(true);
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setTeamError({
      key: _interfaces.TeamStateKeys.RESEND_INVITATION_LINK,
      value: e.message
    }));
    callback == null ? void 0 : callback(null, e.message);
  }

  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.RESEND_INVITATION_LINK,
    value: false
  }));
}

function* resendInvitationEmail({
  payload
}) {
  const {
    callback
  } = payload,
        body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded10);
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.RESEND_INVITATION_LINK,
    value: body.email
  }));

  try {
    yield (0, _effects.call)(_restApi.api.auth.resendInvitationEmail, body);
    callback == null ? void 0 : callback(true);
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setTeamError({
      key: _interfaces.TeamStateKeys.RESEND_INVITATION_LINK,
      value: e.message
    }));
    callback == null ? void 0 : callback(null, e.message);
  }

  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.RESEND_INVITATION_LINK,
    value: false
  }));
}

function* resendInvitationLinkToAllSubTenants({
  payload
}) {
  const {
    callback
  } = payload,
        body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded11);
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.RESEND_INVITATION_LINK,
    value: body.email
  }));

  try {
    yield (0, _effects.call)(_restApi.api.teams.resendInvitationLinkToAllTenants, body);
    callback == null ? void 0 : callback(true);
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setTeamError({
      key: _interfaces.TeamStateKeys.RESEND_INVITATION_LINK,
      value: e.message
    }));
    callback == null ? void 0 : callback(null, e.message);
  }

  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.RESEND_INVITATION_LINK,
    value: false
  }));
}

function* getInvitationLinkConfig() {
  yield (0, _effects.put)(_reducer.actions.setTeamError({
    key: _interfaces.TeamStateKeys.CONFIG_TOKEN_LINK,
    value: false
  }));

  try {
    const invitationLinkConfig = yield (0, _effects.call)(_restApi.api.teams.getInviteLinkConfiguration);
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      inviteTokenState: (0, _extends2.default)({}, invitationLinkConfig)
    }));
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setTeamError({
      key: _interfaces.TeamStateKeys.CONFIG_TOKEN_LINK,
      value: e.message
    }));
  }
}

function* getInvitationLink() {
  yield (0, _effects.put)(_reducer.actions.setTeamError({
    key: _interfaces.TeamStateKeys.GET_TOKEN_LINK,
    value: false
  }));

  try {
    yield (0, _effects.call)(getInvitationLinkConfig);
    const data = yield (0, _effects.call)(_restApi.api.teams.getInviteUserLink);
    const {
      inviteTokenState
    } = yield selectTeamState();
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      inviteTokenState: (0, _extends2.default)({}, inviteTokenState, data)
    }));
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setTeamError({
      key: _interfaces.TeamStateKeys.GET_TOKEN_LINK,
      value: e.message
    }));
  }
}

function* createInvitationLink({
  payload: {
    callback
  }
}) {
  yield (0, _effects.put)(_reducer.actions.setTeamError({
    key: _interfaces.TeamStateKeys.CREATE_TOKEN_LINK,
    value: false
  }));
  const {
    inviteTokenState
  } = yield selectTeamState();

  try {
    const data = yield (0, _effects.call)(_restApi.api.teams.createInviteUserLink, {
      expiresInMinutes: 43200
    });
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      inviteTokenState: (0, _extends2.default)({}, inviteTokenState, data)
    }));
    callback == null ? void 0 : callback(data.token);
  } catch (e) {
    callback == null ? void 0 : callback(null, e.message);
    yield (0, _effects.put)(_reducer.actions.setTeamError({
      key: _interfaces.TeamStateKeys.CREATE_TOKEN_LINK,
      value: e.message
    }));
  }
}

function* updateInvitationLink({
  payload: {
    callback,
    expiresInMinutes,
    shouldSendEmail
  }
}) {
  const {
    inviteTokenState
  } = yield selectTeamState();
  yield (0, _effects.put)(_reducer.actions.setTeamError({
    key: _interfaces.TeamStateKeys.UPDATE_TOKEN_LINK,
    value: false
  }));

  try {
    const data = yield (0, _effects.call)(_restApi.api.teams.updateInviteUserLink, {
      expiresInMinutes,
      shouldSendEmail
    });
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      inviteTokenState: (0, _extends2.default)({}, inviteTokenState, data)
    }));
    callback == null ? void 0 : callback(true);
  } catch (e) {
    callback == null ? void 0 : callback(null, e.message);
    yield (0, _effects.put)(_reducer.actions.setTeamError({
      key: _interfaces.TeamStateKeys.UPDATE_TOKEN_LINK,
      value: e.message
    }));
  }
}

function* deleteInvitationLink({
  payload
}) {
  const {
    callback
  } = payload != null ? payload : {};
  yield (0, _effects.put)(_reducer.actions.setTeamError({
    key: _interfaces.TeamStateKeys.DELETE_TOKEN_LINK,
    value: false
  }));

  try {
    yield (0, _effects.call)(_restApi.api.teams.deleteInviteUserLink);
    yield (0, _effects.put)(_reducer.actions.setTeamState({
      inviteTokenState: undefined
    }));
    callback == null ? void 0 : callback(true);
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setTeamError({
      key: _interfaces.TeamStateKeys.DELETE_TOKEN_LINK,
      value: e.message
    }));
    callback == null ? void 0 : callback(false, e.message);
  }
}

function* openAddUserDialog({
  payload
}) {
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    addUserDialogState: (0, _extends2.default)({
      open: true,
      loading: false,
      error: false
    }, payload)
  }));
}

function* closeAddUserDialog({
  payload
}) {
  const teamState = yield selectTeamState();
  const {
    addUserDialogState: {
      onClose
    }
  } = teamState;
  onClose == null ? void 0 : onClose(payload);
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    addUserDialogState: {
      loading: false,
      error: false,
      open: false
    }
  }));
}

function* openDeleteUserDialog({
  payload
}) {
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    deleteUserDialogState: (0, _extends2.default)({
      open: true,
      loading: false,
      error: false
    }, payload)
  }));
}

function* closeDeleteUserDialog({
  payload
}) {
  const teamState = yield selectTeamState();
  const {
    deleteUserDialogState: {
      onClose
    }
  } = teamState;
  onClose == null ? void 0 : onClose(payload);
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    deleteUserDialogState: {
      loading: false,
      error: false,
      open: false
    }
  }));
}

function* teamSagas() {
  yield (0, _effects.takeLatest)(_reducer.actions.loadUsers, loadUsers);
  yield (0, _effects.takeLatest)(_reducer.actions.loadAllSubTenantsUsers, loadAllSubTenantsUsers);
  yield (0, _effects.takeLatest)(_reducer.actions.loadRoles, loadRoles);
  yield (0, _effects.takeEvery)(_reducer.actions.addUser, addUser);
  yield (0, _effects.takeEvery)(_reducer.actions.addUsersBulk, addUsersBulk);
  yield (0, _effects.takeEvery)(_reducer.actions.addUserToSubTenants, addUserToSubTenants);
  yield (0, _effects.takeEvery)(_reducer.actions.updateUser, updateUser);
  yield (0, _effects.takeEvery)(_reducer.actions.setUserRolesForSubTenants, setUserRolesForSubTenants);
  yield (0, _effects.takeEvery)(_reducer.actions.deleteUser, deleteUser);
  yield (0, _effects.takeEvery)(_reducer.actions.deleteUserFromSubTenants, deleteUserFromSubTenants);
  yield (0, _effects.takeEvery)(_reducer.actions.resendActivationLink, resendActivationLink);
  yield (0, _effects.takeEvery)(_reducer.actions.resendInvitationLink, resendInvitationLink);
  yield (0, _effects.takeEvery)(_reducer.actions.resendInvitationEmail, resendInvitationEmail);
  yield (0, _effects.takeEvery)(_reducer.actions.resendInvitationLinkToAllSubTenants, resendInvitationLinkToAllSubTenants);
  yield (0, _effects.takeEvery)(_reducer.actions.getInvitationLink, getInvitationLink);
  yield (0, _effects.takeEvery)(_reducer.actions.createInvitationLink, createInvitationLink);
  yield (0, _effects.takeEvery)(_reducer.actions.updateInvitationLink, updateInvitationLink);
  yield (0, _effects.takeEvery)(_reducer.actions.deleteInvitationLink, deleteInvitationLink);
  yield (0, _effects.takeEvery)(_reducer.actions.openAddUserDialog, openAddUserDialog);
  yield (0, _effects.takeEvery)(_reducer.actions.closeAddUserDialog, closeAddUserDialog);
  yield (0, _effects.takeEvery)(_reducer.actions.openDeleteUserDialog, openDeleteUserDialog);
  yield (0, _effects.takeEvery)(_reducer.actions.closeDeleteUserDialog, closeDeleteUserDialog);
}
/*********************************
 *  Preview Sagas
 *********************************/


function* loadUsersMock({
  payload
}) {
  var _payload$pageSize2, _payload$pageOffset2, _payload$filter2, _payload$sort2;

  const {
    silentLoading,
    callback
  } = payload;
  const teamState = yield selectTeamState();
  const pageSize = (_payload$pageSize2 = payload.pageSize) != null ? _payload$pageSize2 : teamState.pageSize;
  const pageOffset = (_payload$pageOffset2 = payload.pageOffset) != null ? _payload$pageOffset2 : teamState.pageOffset;
  const filter = (_payload$filter2 = payload.filter) != null ? _payload$filter2 : teamState.filter;
  const sort = (_payload$sort2 = payload.sort) != null ? _payload$sort2 : teamState.sort;
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.USERS,
    value: !silentLoading
  }));
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    pageSize,
    pageOffset,
    filter,
    sort
  }));
  const totalPages = 2;
  const totalItems = 10;
  yield (0, _utils.delay)();
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    users: _dummy.usersDemo,
    totalPages,
    totalItems,
    roles: _dummy.rolesDemo,
    permissions: _dummy.permissionsDemo
  }));
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.USERS,
    value: false
  }));
  callback == null ? void 0 : callback(_dummy.usersDemo);
}

function* loadAllSubTenantsUsersMock({
  payload
}) {
  var _payload$_limit2, _payload$_offset2, _payload$_filter2, _payload$_sortBy2, _payload$_order2;

  const {
    silentLoading,
    callback
  } = payload;
  const teamState = yield selectTeamState();

  const _limit = (_payload$_limit2 = payload._limit) != null ? _payload$_limit2 : teamState.allUsersQueryParams._limit;

  const _offset = (_payload$_offset2 = payload._offset) != null ? _payload$_offset2 : teamState.allUsersQueryParams._offset;

  const _filter = (_payload$_filter2 = payload._filter) != null ? _payload$_filter2 : teamState.allUsersQueryParams._filter;

  const _sortBy = (_payload$_sortBy2 = payload._sortBy) != null ? _payload$_sortBy2 : teamState.allUsersQueryParams._sortBy;

  const _order = (_payload$_order2 = payload._order) != null ? _payload$_order2 : teamState.allUsersQueryParams._order;

  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.USERS,
    value: !silentLoading
  }));
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    allUsersQueryParams: {
      _limit,
      _offset,
      _filter,
      _sortBy,
      _order
    }
  }));
  const totalPages = 2;
  const totalItems = 10;
  yield (0, _utils.delay)();
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    allUsers: _dummy.allUsersDemo,
    totalPages,
    totalItems,
    roles: _dummy.rolesDemo,
    permissions: _dummy.permissionsDemo
  }));
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.USERS,
    value: false
  }));
  callback == null ? void 0 : callback(_dummy.allUsersDemo);
}

function* loadRolesMock({
  payload
}) {
  var _payload$callback3;

  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.ROLES_AND_PERMISSIONS,
    value: true
  }));
  yield (0, _utils.delay)();
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    roles: _dummy.rolesDemo,
    permissions: _dummy.permissionsDemo
  }));
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.ROLES_AND_PERMISSIONS,
    value: true
  }));
  payload == null ? void 0 : (_payload$callback3 = payload.callback) == null ? void 0 : _payload$callback3.call(payload, {
    roles: _dummy.rolesDemo,
    permissions: _dummy.permissionsDemo
  });
}

function* addUserMock({
  payload
}) {
  const {
    callback
  } = payload,
        body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded12);
  const teamState = yield selectTeamState();
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    addUserDialogState: (0, _extends2.default)({}, teamState.addUserDialogState, {
      loading: true
    })
  }));
  yield (0, _utils.delay)();
  const newUser = (0, _extends2.default)({}, _dummy.userTeamDemo, body, {
    id: `${(0, _uuid.v4)()}`
  });
  callback == null ? void 0 : callback(newUser);
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    users: [newUser, ...teamState.users],
    addUserDialogState: {
      open: false,
      loading: false
    }
  }));
}

function* addUserToSubTenantsMock({
  payload
}) {
  const {
    callback
  } = payload,
        body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded13);
  const teamState = yield selectTeamState();
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    addUserDialogState: (0, _extends2.default)({}, teamState.addUserDialogState, {
      loading: true
    })
  }));
  yield (0, _utils.delay)();
  const newUser = (0, _extends2.default)({}, _dummy.userTeamDemo, body, {
    id: `${(0, _uuid.v4)()}`
  });
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    users: [newUser, ...teamState.users],
    addUserDialogState: {
      open: false,
      loading: false
    }
  }));
  callback == null ? void 0 : callback(null);
}

function* updateUserMock({
  payload
}) {
  var _body$roleIds2;

  const {
    callback
  } = payload,
        body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded14);
  const {
    id: userId
  } = body;
  const teamState = yield selectTeamState();
  const oldUserData = teamState.users.find(user => user.id === body.id);
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.UPDATE_USER,
    value: userId || ''
  }));
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    addUserDialogState: (0, _extends2.default)({}, teamState.addUserDialogState, {
      loading: true
    })
  }));
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    addUserDialogState: (0, _extends2.default)({}, teamState.addUserDialogState, {
      loading: true
    }),
    users: teamState.users.map(user => {
      if (user.id === body.id) {
        return (0, _extends2.default)({}, user, body);
      }

      return user;
    })
  }));

  if (oldUserData.roleIds.length > 0 && ((_body$roleIds2 = body.roleIds) == null ? void 0 : _body$roleIds2.length) === 0) {
    body.roleIds = [''];
  }

  yield (0, _utils.delay)();
  const newUser = (0, _extends2.default)({}, oldUserData, body);
  callback == null ? void 0 : callback(newUser);
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    users: teamState.users.map(user => user.id === newUser.id ? (0, _extends2.default)({}, user, newUser, {
      createdAt: user.createdAt,
      customData: user.customData,
      lastLogin: user.lastLogin
    }) : user)
  }));
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.UPDATE_USER,
    value: false
  }));
}

function* deleteUserMock({
  payload
}) {
  const {
    callback
  } = payload,
        body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded15);
  const teamState = yield selectTeamState();
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    deleteUserDialogState: (0, _extends2.default)({}, teamState.deleteUserDialogState, {
      loading: true
    })
  }));
  yield (0, _utils.delay)();
  callback == null ? void 0 : callback(true);
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    users: teamState.users.filter(user => user.id !== body.userId),
    deleteUserDialogState: {
      open: false,
      loading: false
    }
  }));
}

function* deleteUserFromSubTenantsMock({
  payload
}) {
  const {
    callback
  } = payload,
        body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded16);
  const teamState = yield selectTeamState();
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    deleteUserDialogState: (0, _extends2.default)({}, teamState.deleteUserDialogState, {
      loading: true
    })
  }));
  yield (0, _utils.delay)();
  callback == null ? void 0 : callback(true);
  yield (0, _effects.put)(_reducer.actions.setTeamState({
    allUsers: teamState.allUsers.filter(user => user.id !== body.userId),
    deleteUserDialogState: {
      open: false,
      loading: false
    }
  }));
}

function* resendActivationLinkMock({
  payload
}) {
  const {
    callback
  } = payload,
        body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded17);
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.RESEND_ACTIVATE_LINK,
    value: body.userId
  }));
  yield (0, _utils.delay)();
  callback == null ? void 0 : callback(true);
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.RESEND_ACTIVATE_LINK,
    value: false
  }));
}

function* resendInvitationLinkMock({
  payload
}) {
  const {
    callback
  } = payload,
        body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded18);
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.RESEND_INVITATION_LINK,
    value: body.email
  }));
  yield (0, _utils.delay)();
  callback == null ? void 0 : callback(true);
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.RESEND_INVITATION_LINK,
    value: false
  }));
}

function* resendInvitationEmailMock({
  payload
}) {
  const {
    callback
  } = payload,
        body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded19);
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.RESEND_INVITATION_LINK,
    value: body.email
  }));
  yield (0, _utils.delay)();
  callback == null ? void 0 : callback(true);
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.RESEND_INVITATION_LINK,
    value: false
  }));
}

function* resendInvitationLinkToAllSubTenantsMock({
  payload
}) {
  const {
    callback
  } = payload,
        body = (0, _objectWithoutPropertiesLoose2.default)(payload, _excluded20);
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.RESEND_INVITATION_LINK,
    value: body.email
  }));
  yield (0, _utils.delay)();
  callback == null ? void 0 : callback(true);
  yield (0, _effects.put)(_reducer.actions.setTeamLoader({
    key: _interfaces.TeamStateKeys.RESEND_INVITATION_LINK,
    value: false
  }));
}

function* teamSagasMock() {
  yield (0, _effects.takeLatest)(_reducer.actions.loadUsers, loadUsersMock);
  yield (0, _effects.takeLatest)(_reducer.actions.loadAllSubTenantsUsers, loadAllSubTenantsUsersMock);
  yield (0, _effects.takeLatest)(_reducer.actions.loadRoles, loadRolesMock);
  yield (0, _effects.takeEvery)(_reducer.actions.addUser, addUserMock);
  yield (0, _effects.takeEvery)(_reducer.actions.addUserToSubTenants, addUserToSubTenantsMock);
  yield (0, _effects.takeEvery)(_reducer.actions.updateUser, updateUserMock);
  yield (0, _effects.takeEvery)(_reducer.actions.deleteUser, deleteUserMock);
  yield (0, _effects.takeEvery)(_reducer.actions.deleteUserFromSubTenants, deleteUserFromSubTenantsMock);
  yield (0, _effects.takeEvery)(_reducer.actions.resendActivationLink, resendActivationLinkMock);
  yield (0, _effects.takeEvery)(_reducer.actions.resendInvitationLink, resendInvitationLinkMock);
  yield (0, _effects.takeEvery)(_reducer.actions.resendInvitationEmail, resendInvitationEmailMock);
  yield (0, _effects.takeEvery)(_reducer.actions.resendInvitationLinkToAllSubTenants, resendInvitationLinkToAllSubTenantsMock);
  yield (0, _effects.takeEvery)(_reducer.actions.openAddUserDialog, openAddUserDialog);
  yield (0, _effects.takeEvery)(_reducer.actions.closeAddUserDialog, closeAddUserDialog);
  yield (0, _effects.takeEvery)(_reducer.actions.openDeleteUserDialog, openDeleteUserDialog);
  yield (0, _effects.takeEvery)(_reducer.actions.closeDeleteUserDialog, closeDeleteUserDialog);
}
