// @frontegg/react-hooks/node/auth/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  useAuth: true,
  useAuthActions: true,
  useOnRedirectTo: true,
  useAuthRoutes: true,
  useIsAuthenticated: true,
  useAuthUser: true,
  useAuthUserOrNull: true,
  useLoginWithRedirect: true
};
Object.defineProperty(exports, "useAuth", {
  enumerable: true,
  get: function () {
    return _hooks.useAuth;
  }
});
Object.defineProperty(exports, "useAuthActions", {
  enumerable: true,
  get: function () {
    return _hooks.useAuthActions;
  }
});
Object.defineProperty(exports, "useAuthRoutes", {
  enumerable: true,
  get: function () {
    return _hooks.useAuthRoutes;
  }
});
Object.defineProperty(exports, "useAuthUser", {
  enumerable: true,
  get: function () {
    return _hooks.useAuthUser;
  }
});
Object.defineProperty(exports, "useAuthUserOrNull", {
  enumerable: true,
  get: function () {
    return _hooks.useAuthUserOrNull;
  }
});
Object.defineProperty(exports, "useIsAuthenticated", {
  enumerable: true,
  get: function () {
    return _hooks.useIsAuthenticated;
  }
});
Object.defineProperty(exports, "useLoginWithRedirect", {
  enumerable: true,
  get: function () {
    return _hooks.useLoginWithRedirect;
  }
});
Object.defineProperty(exports, "useOnRedirectTo", {
  enumerable: true,
  get: function () {
    return _hooks.useOnRedirectTo;
  }
});

var _hooks = require("./hooks");

var _acceptInvitation = require("./acceptInvitation");

Object.keys(_acceptInvitation).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _acceptInvitation[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _acceptInvitation[key];
    }
  });
});

var _activateAccount = require("./activateAccount");

Object.keys(_activateAccount).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _activateAccount[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _activateAccount[key];
    }
  });
});

var _apiTokens = require("./apiTokens");

Object.keys(_apiTokens).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _apiTokens[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _apiTokens[key];
    }
  });
});

var _forgotPassword = require("./forgotPassword");

Object.keys(_forgotPassword).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _forgotPassword[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _forgotPassword[key];
    }
  });
});

var _resetPhoneNumber = require("./resetPhoneNumber");

Object.keys(_resetPhoneNumber).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _resetPhoneNumber[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _resetPhoneNumber[key];
    }
  });
});

var _login = require("./login");

Object.keys(_login).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _login[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _login[key];
    }
  });
});

var _mfa = require("./mfa");

Object.keys(_mfa).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _mfa[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _mfa[key];
    }
  });
});

var _profile = require("./profile");

Object.keys(_profile).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _profile[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _profile[key];
    }
  });
});

var _signup = require("./signup");

Object.keys(_signup).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _signup[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _signup[key];
    }
  });
});

var _socialLogin = require("./socialLogin");

Object.keys(_socialLogin).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _socialLogin[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _socialLogin[key];
    }
  });
});

var _team = require("./team");

Object.keys(_team).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _team[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _team[key];
    }
  });
});

var _sso = require("./sso");

Object.keys(_sso).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _sso[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _sso[key];
    }
  });
});

var _securityPolicy = require("./securityPolicy");

Object.keys(_securityPolicy).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _securityPolicy[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _securityPolicy[key];
    }
  });
});

var _tenants = require("./tenants");

Object.keys(_tenants).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _tenants[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _tenants[key];
    }
  });
});

var _roles = require("./roles");

Object.keys(_roles).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _roles[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _roles[key];
    }
  });
});

var _sessions = require("./sessions");

Object.keys(_sessions).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _sessions[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _sessions[key];
    }
  });
});

var _sessionsPolicy = require("./sessionsPolicy");

Object.keys(_sessionsPolicy).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _sessionsPolicy[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _sessionsPolicy[key];
    }
  });
});

var _restrictions = require("./restrictions");

Object.keys(_restrictions).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _restrictions[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _restrictions[key];
    }
  });
});

var _provisioning = require("./provisioning");

Object.keys(_provisioning).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _provisioning[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _provisioning[key];
    }
  });
});

var _impersonate = require("./impersonate");

Object.keys(_impersonate).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _impersonate[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _impersonate[key];
    }
  });
});
