// @frontegg/rest-api/node/auth/interfaces.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  SecondaryAuthStrategy: true,
  WebAuthnDeviceType: true,
  MFAStrategyEnum: true
};
exports.WebAuthnDeviceType = exports.SecondaryAuthStrategy = exports.MFAStrategyEnum = void 0;

var _interfaces = require("./secutiry-poilicy/interfaces");

Object.keys(_interfaces).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interfaces[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interfaces[key];
    }
  });
});
let SecondaryAuthStrategy;
exports.SecondaryAuthStrategy = SecondaryAuthStrategy;

(function (SecondaryAuthStrategy) {
  SecondaryAuthStrategy["WebAuthnPlatform"] = "WebAuthnPlatform";
  SecondaryAuthStrategy["WebAuthnCrossPlatform"] = "WebAuthnCrossPlatform";
  SecondaryAuthStrategy["SmsCode"] = "SmsCode";
})(SecondaryAuthStrategy || (exports.SecondaryAuthStrategy = SecondaryAuthStrategy = {}));

let WebAuthnDeviceType;
exports.WebAuthnDeviceType = WebAuthnDeviceType;

(function (WebAuthnDeviceType) {
  WebAuthnDeviceType["Platform"] = "Platform";
  WebAuthnDeviceType["CrossPlatform"] = "CrossPlatform";
})(WebAuthnDeviceType || (exports.WebAuthnDeviceType = WebAuthnDeviceType = {}));

let MFAStrategyEnum;
exports.MFAStrategyEnum = MFAStrategyEnum;

(function (MFAStrategyEnum) {
  MFAStrategyEnum["AuthenticatorApp"] = "AuthenticatorApp";
  MFAStrategyEnum["WebAuthnPlatform"] = "WebAuthnPlatform";
  MFAStrategyEnum["WebAuthnCrossPlatform"] = "WebAuthnCrossPlatform";
  MFAStrategyEnum["SMS"] = "SMS";
})(MFAStrategyEnum || (exports.MFAStrategyEnum = MFAStrategyEnum = {}));
