// @frontegg/rest-api/node/routers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.fronteggUsersUrl = exports.fronteggTenantsUrl = exports.fronteggSilentRefreshTokenUrl = exports.fronteggRefreshTokenUrl = exports.fronteggLogoutApiRoutes = exports.fronteggAuthApiRoutes = void 0;

var _constants = require("./constants");

const fronteggLogoutApiRoutes = [`${_constants.urls.identity.auth.v1}/logout`, `${_constants.urls.oauth.v1}/logout`];
exports.fronteggLogoutApiRoutes = fronteggLogoutApiRoutes;
const fronteggAuthApiRoutes = [...fronteggLogoutApiRoutes, `${_constants.urls.identity.auth.v1}/user/saml/postlogin`, `${_constants.urls.identity.auth.v2}/user/oidc/postlogin`, `${_constants.urls.identity.auth.v1}/user`, `${_constants.urls.identity.auth.v1}/user/mfa/verify`, `${_constants.urls.identity.auth.v1}/user/token/refresh`, `${_constants.urls.identity.users.v1}`, `${_constants.urls.identity.users.v1}/signUp`, `${_constants.urls.identity.auth.v1}/passwordless/magiclink/postlogin`, `${_constants.urls.identity.auth.v1}/passwordless/code/postlogin`, `${_constants.urls.identity.auth.v1}/passwordless/smscode/postlogin`, `${_constants.urls.identity.users.v1}/activate`, `${_constants.urls.identity.users.v1}/invitation/accept`, `${_constants.urls.oauth.v1}/token`, `${_constants.urls.oauth.v1}/authorize/silent`];
exports.fronteggAuthApiRoutes = fronteggAuthApiRoutes;
const fronteggRefreshTokenUrl = `${_constants.urls.identity.auth.v1}/user/token/refresh`;
exports.fronteggRefreshTokenUrl = fronteggRefreshTokenUrl;
const fronteggSilentRefreshTokenUrl = `${_constants.urls.oauth.v1}/authorize/silent`;
exports.fronteggSilentRefreshTokenUrl = fronteggSilentRefreshTokenUrl;
const fronteggUsersUrl = `${_constants.urls.identity.users.v2}/me`;
exports.fronteggUsersUrl = fronteggUsersUrl;
const fronteggTenantsUrl = `${_constants.urls.identity.users.v2}/me/tenants`;
exports.fronteggTenantsUrl = fronteggTenantsUrl;
