// @frontegg/rest-api/node/sub-tenants/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.addUserRolesForSubTenants = addUserRolesForSubTenants;
exports.addUserToTenantAndSubTenants = addUserToTenantAndSubTenants;
exports.loadAllUsers = loadAllUsers;
exports.removeUserFromTenantAndSubTenants = removeUserFromTenantAndSubTenants;
exports.removeUserRolesFromSubTenants = removeUserRolesFromSubTenants;
exports.setUserRolesForSubTenants = setUserRolesForSubTenants;

var _fetch = require("../fetch");

var _constants = require("../constants");

async function loadAllUsers(params) {
  return (0, _fetch.Get)(_constants.urls.identity.subTenants.v1, params);
}

async function addUserToTenantAndSubTenants(body) {
  return (0, _fetch.Post)(_constants.urls.identity.subTenants.v1, body);
}

async function removeUserFromTenantAndSubTenants(body) {
  return (0, _fetch.Delete)(_constants.urls.identity.subTenants.v1, body);
}

async function addUserRolesForSubTenants(userId, body) {
  return (0, _fetch.Post)(`${_constants.urls.identity.subTenants.v1}/${userId}/roles`, body);
}

async function removeUserRolesFromSubTenants(userId, body) {
  return (0, _fetch.Delete)(`${_constants.urls.identity.subTenants.v1}/${userId}/roles`, body);
}

async function setUserRolesForSubTenants(userId, body) {
  return (0, _fetch.Patch)(`${_constants.urls.identity.subTenants.v1}/${userId}/roles`, body);
}
